### DISCUSSION TITLE
StateNotifier using either Ref or WidgetRef

### DISCUSSION THREAD

**Discussion by scopendo** [rrousselGit/riverpod#1152] [Q&A] [answered]
This discussion is about having a state notifier that wants to read values held by other providers, using either the `Ref` passed to a provider's create function or the `WidgetRef` of a `ConsumerWidget`.  The reason is that sometimes the (generic) state notifier is instantiated via the provider create function and sometimes it is set via a `ProviderScope` override.

The problem encountered is that `Ref` and `WidgetRef` do not appear to implement a common interface.  I got around this by passing the `Reader` method to the state notifier instead.

```dart
abstract class AppController<State> extends StateNotifier<State> {
  AppController(this.read, State initialState) : super(initialState);

  final Reader read;
}
```

However, I wondered why `Ref` and `WidgetRef` don't share a common interface? It's possible that a state notifier might also want to listen to or refresh other providers.

**Comment by rrousselGit** [comment, marked as answer]
> However, I wondered why Ref and WidgetRef don't share a common interface?

That's voluntary. The fact that you need a common interface shows a flaw in your architecture

You most definitely did something wrong somewhere. You should never need to use WidgetRef outside of widgets

**Reply by rrousselGit** [reply]
In particular, you should look into Inversion of Control.

Chances are there's a way to refactor the code path currently using WidgetRef and extract that logic into a provider instead, therefore using Ref.

**Reply by scopendo** [reply]
I'm happy to accept that but at the moment the only way I see around it is to pass the dependencies in when overriding the provider's value on the provider scope, rather than using a ref internally.

Suppose that we have a list of "categories" and that tapping on one of them opens the category screen, with the selected or "current" category being set on a ProviderScope, but not as a simple state value but rather wrapped in a StateNotifier so that we can perform "someAction" on the category screen that changes the current category state.

```dart
class _CategoryListTile extends ConsumerWidget {
  const _CategoryListTile({required this.category});

  final CategoryData category;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return ListTile(
      title: Text(category.name),
      onTap: () => Navigator.push(
        context,
        MaterialPageRoute<void>(
          builder: (_) => ProviderScope(
            overrides: [
              currentCategoryProvider.overrideWithValue(
                CategoryController(ref.read, category), // <-- where I was seemingly doing the wrong thing
              ),
            ],
            child: const CategoryScreen(),
          ),
        ),
      ),
    );
  }
}

final currentCategoryProvider = StateNotifierProvider<CategoryController, CategoryData>(
  (_) => throw 'Provider value must be overridden',
);

class CategoryController extends StateNotifier<CategoryData> {
  CategoryController(this.read, CategoryData category) : super(category);

  final Reader read;

  Future<void> someAction() async {
    final dependency1 = read(dependency1Provider);
    final dependency2 = read(dependency2Provider);
    ...
    state = state.copyWith(...);
   }
}
```

In this scenario, I don't think I'm misusing Riverpod by accessing the dependencies internally, rather than passing them in.  I also don't see how this CategoryController could source its reader function from anywhere but a WidgetRef given that the category controller value is set directly on the ProviderScope via an override.

**Reply by rrousselGit** [reply]
Try using overrideWithProvider insted of overrideWithValue

```dart
currentCategoryProvider.overrideWithProvider(
  StateNotifierProvider((ref) => CategoryController(ref, category),
)
```

**Reply by scopendo** [reply]
Perfect.
